lib/ansible/modules/system/zfs_permissions.py

```diff
@@ -0,0 +1,271 @@
+        self.recursive = module.boolean(module.params.get('recursive'))
+        self.scope = None
+        self.changed = False
+        self.__current_perms = None
+
+        if self.state == 'present' and not self.perms:
+            self.module.fail_json(msg='The `permissions` option is required for state=present')
+
+        if self.state == 'present' and not (self.users or self.groups or self.everyone):
+            self.module.fail_json(msg='One of `users`, `groups`, or `everyone` must be set')
+
+        for splittable in ('users', 'groups', 'perms'):
+            if getattr(self, splittable):
+                setattr(self, splittable, getattr(self, splittable).split(','))
+
+        local = module.boolean(module.params.get('local'))
+        descendents = module.boolean(module.params.get('descendents'))
+        if (local and descendents) or (not local and not descendents):
+            self.scope = 'ld'
+        elif local:
+            self.scope = 'l'
+        elif descendents:
+            self.scope = 'd'
+        else:
+            self.module.fail_json(msg='Impossible value for local and descendents')
+
+        self.subcommand = 'allow'
+        self.recursive_opt = []
+        if self.state == 'absent':
+            self.subcommand = 'unallow'
+            if self.recursive:
+                self.recursive_opt = ['-r']
+            else:
+                self.recursive_opt = []
+
+        self.run()
+
+    @property
+    def current_perms(self):
+        if self.__current_perms is None:
+            rc, out, err = self.run_command(['zfs', 'allow', self.name])
+            if rc:
+                self.module.fail_json(msg='Getting permissions for %s failed: %s' % (self.name, err))
+            perms = dict(l  = dict(u=dict(), g=dict(), e=[]),
+                         d  = dict(u=dict(), g=dict(), e=[]),
+                         ld = dict(u=dict(), g=dict(), e=[]))
+            reading = None
+            for line in out.splitlines():
+                if line == 'Local permissions:':
+                    reading = 'l'
+                elif line == 'Descendent permissions:':
+                    reading = 'd'
+                elif line == 'Local+Descendent permissions:':
+                    reading = 'ld'
+                elif line.startswith('\tuser '):
+                    user, cur_perms = line.split()[1:3]
+                    perms[reading]['u'][user] = cur_perms.split(',')
+                elif line.startswith('\tgroup '):
+                    group, cur_perms = line.split()[1:3]
+                    perms[reading]['g'][group] = cur_perms.split(',')
+                elif line.startswith('\teveryone '):
+                    perms[reading]['e'] = line.split()[1].split(',')
+            self.__current_perms = perms
+        return self.__current_perms
+
+    def run_command(self, cmd):
+        progname = cmd[0]
+        cmd[0] = self.module.get_bin_path(progname, True)
```

Comment:
You are looking for zfs(1M) everytime you run this function. Why not do it in __init__ when initiating class and using a variable, e.g. self._zfs_binary?